includes/storage/SQLStore/SMW_SQLStore3_Queries.php

```diff
@@ -93,113 +86,39 @@ public function __construct( SMWSQLStore3 $parentstore, $dbslave ) {
 	 * @return array of error strings (empty if no errors occurred)
 	 */
 	public function refreshConceptCache( Title $concept ) {
-		global $smwgQMaxLimit, $smwgQConceptFeatures, $wgDBtype;
-
-		$fname = 'SMW::refreshConceptCache';
-
-		$db = $this->store->getConnection();
+		$conceptCache = new ConceptCache( $this, $this->store );
 
-		$cid = $this->store->smwIds->getSMWPageID( $concept->getDBkey(), SMW_NS_CONCEPT, '', '' );
-		$cid_c = $this->store->smwIds->getSMWPageID( $concept->getDBkey(), SMW_NS_CONCEPT, '', '', false );
+		$errors = $conceptCache->refresh( $concept );
 
-		if ( $cid != $cid_c ) {
-			$this->errors[] = "Skipping redirect concept.";
-			return $this->errors;
-		}
-
-		$values = $this->store->getPropertyValues(
-			SMWDIWikiPage::newFromTitle( $concept ),
-			new SMWDIProperty( '_CONC' )
-		);
-
-		$di = end( $values );
-		$desctxt = ( $di !== false ) ? $di->getConceptQuery() : false;
-		$this->errors = array();
-
-		if ( $desctxt ) { // concept found
-			$this->queryMode = SMWQuery::MODE_INSTANCES;
-			$this->queryParts = array();
-			$this->hierarchyCache = array();
-			$this->executedQueries = array();
-			$this->sortKeys = array();
-			SMWSQLStore3Query::$qnum = 0;
-
-			// Pre-process query:
-			$qp = new SMWQueryParser( $smwgQConceptFeatures );
-			$desc = $qp->getQueryDescription( $desctxt );
-
-			$this->queryBuilder->setSortKeys( $this->sortKeys );
-			$this->queryBuilder->buildQueryContainer( $desc );
+		$this->cleanUp();
 
-			$qid = $this->queryBuilder->getLastContainerId();
-			$this->queryParts = $this->queryBuilder->getQueryContainer();
-			$this->errors = $this->queryBuilder->getErrors();
+		return $errors;
+	}
 
-			if ( $qid < 0 ) {
-				return;
-			}
+	/**
+	 * @param string $desctxt
+	 *
+	 * @return SqlQueryPart
+	 */
+	public function doSomeStuffAndReturnSomething( $desctxt ) {
```

Comment:
`$desctxt` contains the raw `[[Modification date::+]]` query and the code within `doSomeStuffAndReturnSomething` is to resolve the query tree and return a sql transformation which is to be used to build the final joins/from/where sql condition.